addons/html_builder/static/src/core/builder_selection_restriction_plugin.js

```diff
@@ -0,0 +1,242 @@
+
+// we probably need to move these special blocks to a resource in different plugins
+const SPECIAL_BLOCK_WITH_TEXT_IN_NON_DIV = ["BLOCKQUOTE"];
+const UNCROSSABLE_ELEMENT_SELECTOR = ["blockquote", "form", "div", "section", ".alert", ".row"];
+
+export class BuilderSelectionRestrictionPlugin extends Plugin {
+    static id = "builderSelectionRestriction";
+    static dependencies = ["history", "selection", "operation", "builderOptions"];
+
+    resources = {
+        uncrossable_element_selector: UNCROSSABLE_ELEMENT_SELECTOR,
+        special_block_with_text_in_non_div: SPECIAL_BLOCK_WITH_TEXT_IN_NON_DIV,
+    };
+
+    setup() {
+        this.uncrossable_element_selector = [
+            ...new Set(this.getResource("uncrossable_element_selector")),
+        ].join(",");
+        this.special_block_with_text_in_non_div = [
+            ...new Set(this.getResource("special_block_with_text_in_non_div")),
+        ];
+        this.selectionCorrected = false;
+
+        this.addDomListener(this.editable, "keydown", this.onKeydown);
+        this.addDomListener(this.document, "mouseup", this.restrictSelectionInClosestDiv);
+        this.addDomListener(this.document, "touchend", this.restrictSelectionInClosestDiv);
+
+        // doing this manually instead of using addDomListener. This is because
+        // addDomListener will ignore all events from protected targets. But in
+        // our case, we still want to update the containers.
+        this.onClick = this.onClick.bind(this);
+        this.editable.addEventListener("click", this.onClick, { capture: true });
+    }
+
+    destroy() {
+        this.editable.removeEventListener("click", this.onClick, { capture: true });
+    }
+
+    onClick(ev) {
+        this.dependencies.operation.next(() => {
+            if (this.selectionCorrected) {
+                this.selectionCorrected = false;
+                return;
+            }
+            this.dependencies.builderOptions.updateContainers(ev.target);
+        });
+    }
+
+    onKeydown(ev) {
+        if (getActiveHotkey(ev) === "control+a") {
+            const { editableSelection, currentSelectionIsInEditable } =
+                this.dependencies.selection.getSelectionData();
+            if (
+                !currentSelectionIsInEditable ||
+                // edge case: when click the image, then ctrl+a the selection should stay on the image
+                editableSelection.commonAncestorContainer.nodeName === "FIGURE" ||
+                // edge case: when main body is empty, and click the footer outer blue container
+                // then ctrl+a, the selection is collapsed in editable but to the main body div,
+                // which causes options updating
+                (editableSelection.anchorNode.isContentEditable === false &&
+                    editableSelection.isCollapsed)
+            ) {
+                ev.preventDefault();
+                ev.stopPropagation();
```

Comment:
Can't you do it directly at the start ? I see that you also do it outside of the if, so better regroup them all and do it once at the beginning :slightly_smiling_face: Is there a reason why the other `stopPropagation` below is this far from `preventDefault` ? :thinking: